### PR TITLE
[Backport] bitcoin/bitcoin#13825 kill accounts

### PR DESCRIPTION
**Pull request by nuttycom**
This is a backport of the following commits:

- from bitcoin/bitcoin#13825
- pick bitcoin/bitcoin@c9c32e6b844fc79467b7e24c6c916142a0d08484
  - from bitcoin/bitcoin#14023
    - pick bitcoin/bitcoin@f0dc850bf698f7377797d7d68365d4fc79b0221c
      - from bitcoin/bitcoin#13566
        - pick bitcoin/bitcoin@702ae1e21a09d8c31406839c4ea507c5fa276898
        - pick bitcoin/bitcoin@cf15761f6d4526d205126fbf5f088ac8edebeb57
        - pick bitcoin/bitcoin@0f3d6e9ab75837ead3acfda342aa3ea404efb81d
        - pick bitcoin/bitcoin@7110c830f8c5de3570178bf4e5d28fe3e4f109e1
        - pick bitcoin/bitcoin@ef7bc8893c7a953953aa457736d79c28a4f45792
        - pick bitcoin/bitcoin@4279da47855ec776f8d57c6579fe89afc9cbe8c1
    - pick bitcoin/bitcoin@c410f415758913c933ad6c71cf50227cc85aa385
    - from bitcoin/bitcoin#9614
      - pick bitcoin/bitcoin@02d9f50d5f3c96fe888c230d59c5afdab4c7c6a3
      - pick bitcoin/bitcoin@82b7dc373afff277d3083d09e68c6b69778f312f
        - from bitcoin/bitcoin#8061
          - pick bitcoin/bitcoin@ecb9741ec3067f67f595126869d0200c62064bbd
  - from bitcoin/bitcoin#6851
    - pick bitcoin/bitcoin@3e7c89196c
  - from bitcoin/bitcoin#8828
    - pick bitcoin/bitcoin@86029e72c96436605f47153c1f09e151e1847c28
      - from bitcoin/bitcoin#8696
        - just the first two commits, as AccountingEntry is going away
        - pick bitcoin/bitcoin@d2e678d7d220e2f430c2cb4bde559acee00c8c7c
        - pick bitcoin/bitcoin@59adc86680911a83dcccf5b23db299a370fc92f2
    - from bitcoin/bitcoin#8028
      - pick bitcoin/bitcoin@0fd599767d2dabf25ac8c3543cb586cfc4b9d816

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -769,24 +499,23 @@ UniValue getbalance(const UniValue& params, bool fHelp)
 
     LOCK2(cs_main, pwalletMain->cs_wallet);
 
-    if (params.size() == 0)
-        return  ValueFromAmount(pwalletMain->GetBalance());
+    const UniValue& dummy_value = params[0];
+    if (!dummy_value.isNull() && dummy_value.get_str() != "*") {
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:502]: This makes sense contextually from upstream, but given that we've always told people in Zcash to hard-code this to the empty string, that should be our default.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -162,11 +152,6 @@ UniValue getnewaddress(const UniValue& params, bool fHelp)
 
     LOCK2(cs_main, pwalletMain->cs_wallet);
 
-    // Parse the account first so we don't generate a key if there's an error
-    string strAccount;
-    if (params.size() > 0)
-        strAccount = AccountFromValue(params[0]);
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:168]: Removing this check while leaving `params.size() > 1` above means that this parameter position can now be set arbitrarily. We should continue to enforce that this is the empty string if present.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -1100,16 +690,13 @@ UniValue addmultisigaddress(const UniValue& params, bool fHelp)
 
     LOCK2(cs_main, pwalletMain->cs_wallet);
 
-    string strAccount;
-    if (params.size() > 2)
-        strAccount = AccountFromValue(params[2]);
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:1169]: Check this is still the empty string.

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -1332,27 +851,36 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
...
-void ListTransactions(const CWalletTx& wtx, const string& strAccount, int nMinDepth, bool fLong, UniValue& ret, const isminefilter& filter)
+/**
+ * List transactions based on the given criteria.
+ *
+ * @param  pwallet    The wallet.
+ * @param  wtx        The wallet transaction.
+ * @param  nMinDepth  The minimum confirmation depth.
+ * @param  fLong      Whether to include the JSON version of the transaction.
+ * @param  ret        The UniValue into which the result is stored.
+ * @param  filter     The "is mine" filter bool.
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:859]: Some of these we don't yet have.

**Comment by nuttycom** [comment]
TODO: Add the change to the release notes. Also, check Zecwallet to ensure none of these are being used.

**Review by str4d — APPROVED** [review]
utACK modulo above comments. Really nice!

**Review by str4d — APPROVED** [review]
re-utACK

**Comment by str4d** [comment]
@zkbot r+

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -3144,16 +3155,17 @@ void CWallet::ResendWalletTransactions(int64_t nBestBlockTime)
...
 
 
-CAmount CWallet::GetBalance() const
+CAmount CWallet::GetBalance(const isminefilter& filter) const
 {
     CAmount nTotal = 0;
     {
         LOCK2(cs_main, cs_wallet);
-        for (map<uint256, CWalletTx>::const_iterator it = mapWallet.begin(); it != mapWallet.end(); ++it)
+        for (const auto& entry : mapWallet)
```
Comment (LarryRuane) [inline src/wallet/wallet.cpp:3109]: I like how you simplified these loops; I think there's one more you could change, in `CWallet::ReorderTransactions()`
```diff
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1418,9 +1418,8 @@ DBErrors CWallet::ReorderTransactions()
     typedef multimap<int64_t, TxPair > TxItems;
     TxItems txByTime;
 
-    for (map<uint256, CWalletTx>::iterator it = mapWallet.begin(); it != mapWallet.end(); ++it)
-    {
-        CWalletTx* wtx = &((*it).second);
+    for (auto& entry : mapWallet) {
+        CWalletTx* wtx = &entry.second;
         txByTime.insert(make_pair(wtx->nTimeReceived, TxPair(wtx, (CAccountingEntry*)0)));
     }
     list<CAccountingEntry> acentries;
```

**Pull request merged**